axonhub-client/src/main/java/io/axoniq/axonhub/client/query/subscription/AxonHubSubscriptionQueryResult.java

```diff
@@ -0,0 +1,154 @@
+import io.axoniq.axonhub.client.AxonHubConfiguration;
+import io.axoniq.axonhub.client.Publisher;
+import io.axoniq.axonhub.client.query.RemoteQueryException;
+import io.axoniq.axonhub.client.util.FlowControllingStreamObserver;
+import io.axoniq.axonhub.grpc.FlowControl;
+import io.grpc.stub.StreamObserver;
+import org.axonframework.common.Registration;
+import org.axonframework.queryhandling.DefaultSubscriptionQueryResult;
+import org.axonframework.queryhandling.QueryResponseMessage;
+import org.axonframework.queryhandling.SubscriptionQueryBackpressure;
+import org.axonframework.queryhandling.SubscriptionQueryMessage;
+import org.axonframework.queryhandling.SubscriptionQueryResult;
+import org.axonframework.queryhandling.SubscriptionQueryUpdateMessage;
+import reactor.core.publisher.EmitterProcessor;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.MonoSink;
+
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static io.axoniq.axonhub.SubscriptionQueryRequest.newBuilder;
+import static java.util.Optional.ofNullable;
+
+/**
+ *
+ * SubscriptionQueryResult that emits initial response and update when subscription query response message is received.
+ *
+ * @author Sara Pellegrini
+ */
+class AxonHubSubscriptionQueryResult implements
+        Supplier<SubscriptionQueryResult<QueryResponse, QueryUpdate>>,
+        StreamObserver<SubscriptionQueryResponse> {
+
+    private final FlowControllingStreamObserver<SubscriptionQueryRequest> requestObserver;
+
+    private final SubscriptionQueryResult<QueryResponse, QueryUpdate> result;
+
+    private final SubscriptionQuery subscriptionQuery;
+
+    private final FluxSink<QueryUpdate> updateMessageFluxSink;
+
+    private final Runnable onDispose;
+
+    private MonoSink<QueryResponse> initialResultSink;
+
+    AxonHubSubscriptionQueryResult(
+            SubscriptionQuery query,
+            Function<StreamObserver<SubscriptionQueryResponse>, StreamObserver<SubscriptionQueryRequest>> openStreamFn,
+            AxonHubConfiguration configuration,
+            SubscriptionQueryBackpressure backPressure,
+            int bufferSize, Runnable onDispose) {
+        this.onDispose = onDispose;
+
+        EmitterProcessor<QueryUpdate> processor = EmitterProcessor.create(bufferSize);
+
+        this.subscriptionQuery = query;
+        this.updateMessageFluxSink = processor.sink(backPressure.getOverflowStrategy());
+
+        StreamObserver<SubscriptionQueryRequest> subscription = openStreamFn.apply(this);
+        Function<FlowControl, SubscriptionQueryRequest> requestMapping = flowControl ->
+                newBuilder().setFlowControl(SubscriptionQuery.newBuilder(subscriptionQuery).setNumberOfPermits(flowControl.getPermits())).build();
+        requestObserver = new FlowControllingStreamObserver<>(subscription, configuration, requestMapping, t -> false);
```

Comment:
Add requestObserver.sendInitialPermits(); to start flow control.

Reply:
thanks, it must be lost during refactoring